src/pages/patientView/sampleManager.tsx

```diff
@@ -18,23 +17,71 @@ export function placeArrow(tooltipEl: any) {
+// 1. based on sample collection data (timeline event)
+// 2. if all cases have derived normalized case types, put primary first
+// 3. natural sort of sample ids
 export function sortSamples(samples: Array<ClinicalDataBySampleId>,
-                            clinicalDataLegacyCleanAndDerived: { [s:string]:any }) {
-    // sort by
-    // 1. based on sample collection data (timeline event) (unimplemented)
-    // 2. if all cases have derived normalized case types, put primary first (unimplemented)
-    // 3. natural sort of sample ids
-    let sortedSampleIDs: string[] = [];
-    sortedSampleIDs = sortedSampleIDs.concat(samples.map((sample) => sample.id)).sort(naturalSort).reverse();
-
-    return _.sortBy(samples, function(sample) {
-        return sortedSampleIDs.indexOf(sample.id);
+                            clinicalDataLegacyCleanAndDerived: { [s:string]:any },
+                            events?: any) {
+    // natural sort
+    let naturalSortedSampleIDs: string[] = [];
+    naturalSortedSampleIDs = naturalSortedSampleIDs.concat(samples.map((sample) => sample.id)).sort(naturalSort);
+
+    // based on sample collection data (timeline event)
+    let collectionDayMap: {[s:string]:number} = {};
+    if (events) {
+        let specimenEvents = events.filter((e: ClinicalEvent) => (e.eventType === 'SPECIMEN'));
+
+        for (let i: number = 0; i < specimenEvents.length; i++) {
+            let sampleAttr = _.find(specimenEvents[i].attributes, (attr: ClinicalEventData) => {
+                // TODO: This is legacy support for old timeline data that does not use SAMPLE_ID, but one of the specrefnum
+                return (attr.key === "SAMPLE_ID" || attr.key === "SpecimenReferenceNumber" || attr.key === "SPECIMEN_REFERENCE_NUMBER") &&
+                    (naturalSortedSampleIDs.indexOf(attr.value) !== -1);
+            });
+            if (sampleAttr && sampleAttr.value !== null && sampleAttr.value !== undefined) {
+                collectionDayMap[sampleAttr.value] = specimenEvents[i].startNumberOfDaysSinceDiagnosis;
+            }
+        }
+    }
+
+    // create new object array, to allow sorting of samples by multiple fields
+    type sampleOrderT = {
+        id: string;
+        // fields to sort by
+        eventOrdering?: number;
+        sampleTypeIndex: number;
+        naturalSortIndex: number;
+    };
+    // put primaries first (could be extended with more if necessary)
+    let sampleTypeOrdering: string[] = ['Primary'];
+    let sampleOrder: sampleOrderT[] = [];
+    
+    for (let i: number = 0; i < samples.length; i++) {
+        let id = samples[i].id;
+        // 1. based on sample collection data (timeline event)
+        let eventOrdering = collectionDayMap[id];
+
+        // 2. if cases have derived normalized case types, put primary first
+        let sampleTypeIndex = sampleTypeOrdering.indexOf(clinicalDataLegacyCleanAndDerived[id].DERIVED_NORMALIZED_CASE_TYPE) * -1;
```

Comment:
ah because `indexOf` returns -1 if it's not found. In this case the's only one item in the sampleTypeOrdering list, so it works. Prolly more clear to handle the -1 case separately.